torchvision/prototype/datasets/_builtin/imagenet.py

```diff
@@ -20,176 +19,176 @@
-
-        return dict(
-            dict(zip(("label", "wnid"), label_data if label_data else (None, None))),
-            path=path,
-            image=EncodedImage.from_file(buffer),
+
+        meta_dp = Mapper(meta_dp, extract_categories_and_wnids)
+        _, wnids = zip(*next(iter(meta_dp)))
+
+        label_dp = LineReader(label_dp, decode=True, return_path=False)
+        label_dp = Mapper(label_dp, imagenet_label_to_wnid)
+        label_dp: IterDataPipe[Tuple[int, str]] = Enumerator(label_dp, 1)
+        label_dp = hint_sharding(label_dp)
+        label_dp = hint_shuffling(label_dp)
+
+        dp = IterKeyZipper(
+            label_dp,
+            images_dp,
+            key_fn=getitem(0),
+            ref_key_fn=path_accessor(val_test_image_key),
+            buffer_size=INFINITE_BUFFER_SIZE,
         )
+        dp = Mapper(dp, prepare_val_data)
 
-    def _datapipe(self, resource_dps: List[IterDataPipe]) -> IterDataPipe[Dict[str, Any]]:
-        if self._split in {"train", "test"}:
-            dp = resource_dps[0]
-
-            # the train archive is a tar of tars
-            if self._split == "train":
-                dp = TarArchiveReader(dp)
-
-            dp = hint_sharding(dp)
-            dp = hint_shuffling(dp)
-            dp = Mapper(dp, self._prepare_train_data if self._split == "train" else self._prepare_test_data)
-        else:  # config.split == "val":
-            images_dp, devkit_dp = resource_dps
-
-            meta_dp, label_dp = Demultiplexer(
-                devkit_dp, 2, self._classifiy_devkit, drop_none=True, buffer_size=INFINITE_BUFFER_SIZE
-            )
-
-            meta_dp = Mapper(meta_dp, self._extract_categories_and_wnids)
-            _, wnids = zip(*next(iter(meta_dp)))
-
-            label_dp = LineReader(label_dp, decode=True, return_path=False)
-            label_dp = Mapper(label_dp, self._imagenet_label_to_wnid)
-            label_dp: IterDataPipe[Tuple[int, str]] = Enumerator(label_dp, 1)
-            label_dp = hint_sharding(label_dp)
-            label_dp = hint_shuffling(label_dp)
-
-            dp = IterKeyZipper(
-                label_dp,
-                images_dp,
-                key_fn=getitem(0),
-                ref_key_fn=path_accessor(self._val_test_image_key),
-                buffer_size=INFINITE_BUFFER_SIZE,
-            )
-            dp = Mapper(dp, self._prepare_val_data)
-
-        return Mapper(dp, self._prepare_sample)
-
-    def __len__(self) -> int:
-        return {
+    dp = Mapper(dp, prepare_sample)
+    return TakerDataPipe(
+        dp,
+        num_take={
             "train": 1_281_167,
             "val": 50_000,
             "test": 100_000,
-        }[self._split]
+        }[split],
+    )
+
 
-    def _generate_categories(self) -> List[Tuple[str, ...]]:
-        self._split = "val"
-        resources = self._resources()
+def generate_categories(root: Union[str, pathlib.Path], **kwargs: Any) -> List[Tuple[str, ...]]:
```

Comment:
This is currently not working. Before, we checked for this method and called it through the already exposed dataset class. Now, we either need to expose this explicitly or find another way to call it. 

In one of the very first iterations I simply added a `if __name__ == "__main__": ...` to the module and performed everything there. This way we wouldn't have an option to generate multiple category files at once, but that is probably not a significant issue.